scrapy/http/request/json_request.py

```diff
@@ -0,0 +1,29 @@
+"""
+This module implements the JSONRequest class which is a more convenient class
+(than Request) to generate JSON Requests.
+
+See documentation in docs/topics/request-response.rst
+"""
+
+import json
+import warnings
+
+from scrapy.http.request import Request
+
+
+class JSONRequest(Request):
+    def __init__(self, *args, **kwargs):
+        body_passed = 'body' in kwargs
+        data = kwargs.pop('data', None)
+        if body_passed and data:
+            warnings.warn('Both body and data passed. data will be ignored')
+
+        elif not body_passed and data:
```

Comment:
I think that if ``data`` is an empty list (which would be false), we still want to send it. I suggest checking both ``body`` and ``data`` for ``None`` - e.g. ``JSONRequest(body=None, data=[])`` should still be valid, I don't think we should distinguish between passing ``None`` and not passing anything (usually it's not done for convenience).

Reply:
Changed